receiver.go

```diff
@@ -103,126 +78,66 @@ func (r *receiveHandler) taskLoop() {
 	}
 }
 
-func (r *receiveHandler) findServeHandler(n string) (*ServeHandlerDescription, WrapErrorFunc, error) {
-	p, m := splitMethodName(n)
-	prot, found := r.protocols[p]
-	if !found {
-		return nil, r.wrapErrorFunc, ProtocolNotFoundError{p}
-	}
-	srv, found := prot.Methods[m]
-	if !found {
-		return nil, r.wrapErrorFunc, MethodNotFoundError{p, m}
-	}
-	return &srv, prot.WrapError, nil
-}
-
-func (d *receiveHandler) Receive(length int) error {
-	var requestType MethodType
-	if err := d.reader.Decode(&requestType); err != nil {
-		return err
-	}
-	handler, ok := d.messageHandlers[requestType]
-	if !ok {
-		return NewDispatcherError("invalid message type")
-	}
-	if length != handler.messageLength {
-		return NewDispatcherError("wrong number of fields for message (got n=%d, expected n=%d)", length, handler.messageLength)
-
+func (r *receiveHandler) Receive(rpc rpcMessage) error {
+	switch rpc.Type() {
+	case MethodNotify:
+		if rpcData, ok := rpc.(*RPCNotifyData); ok {
```

Comment:
i think for all of these cases, we should return an error if !ok. But I think you just do a type switch ( https://golang.org/doc/effective_go.html#type_switch ) and avoid all the boilerplate